### DISCUSSION TITLE
Flat ride with its drawing code in a script file

### DISCUSSION THREAD

**Discussion by frutiemax** [OpenRCT2/OpenRCT2#16154] [Q&A] [answered]
It would be nice in the future to let the ride maker code its own ride drawing code in a script and not in C++. The script file would be embedded in the .parkobj file and would reference image ids that are contain within it. This would prevent hardcoding more and more ride types in the base code, and probably reduce the amount of parameters needed in the .parkobj.

**Comment by janisozaur** [comment]
I'm against this. I don't think running plugins in drawing context would be efficient enough

**Comment by frutiemax** [comment]
I don't think that performance is a valid argument since OpenTTD does its AI with scripts, the same with any modern game engine. The script context wouldn't be at the drawing level, just before it I.e. it would decide the positions of the images, when to play sound, etc.

**Reply by ZehMatt** [reply]
Its is a perfectly reasonable argument. OpenTTD has only a max of 15 companies so simulating the AI via scripting is perfectly fine in this context. Also modern engines usually don't have the performance heavy tasks inside the scripting system, OpenRCT2 is also not something you can compare to most of the games, there are actually not a lot of games where you even get to 10k entities let alone nearly 1 million tile elements. Let alone we have to yet fix the current regression in performance with the new limits.

**Comment by IntelOrca** [comment]
I think it makes sense to see just how much we can define the ride types declaratively. @duncanspumpkin did you have some ideas for this?

**Comment by duncanspumpkin** [comment]
We will likely end up with a good many variations for different flat rides. I think it would make sense to have perhaps some sort of striped down basic script to do the drawing.

**Comment by ZehMatt** [comment, marked as answer]
Before we go into scripting territory we should first try to cleanup our internal structures to be more data oriented. Once that's done all the system has to do is pull of the data of the script.

**Comment by frutiemax** [comment]
I didn't realize that it could be a bottleneck since there can be a lot of rides in 1 park. I think we should continue to encode more fields in the RideTypeDescriptor for the painting.
